### PR TITLE
refactor(web): registry refactored to rest

### PR DESCRIPTION
**Pull request by robot9706**
- Removed registry gRPC (except `RegistryType` enum, it's required by other services) API and added REST API

### CODE REVIEW

File: web/crux-ui/src/models/registry.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable max-classes-per-file */
```
Comment (polaroi8d) [inline web/crux-ui/src/models/registry.ts:1]: We added this to the .eslintrc so you'll don't need to use this in here.

File: web/crux/src/app/registry/registry.dto.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+export const REGISTRY_TYPE_VALUES = ['v2', 'hub', 'gitlab', 'github', 'google', 'unchecked'] as const
+export type RegistryType = (typeof REGISTRY_TYPE_VALUES)[number]
+
+export const GITHUB_NAMESPACE_VALUES = ['organization', 'user'] as const
+export const GITLAB_NAMESPACE_VALUES = ['group', 'project'] as const
+export type GitlabNamespace = (typeof GITLAB_NAMESPACE_VALUES)[number]
+export type GithubNamespace = (typeof GITHUB_NAMESPACE_VALUES)[number]
+export type RegistryNamespace = GitlabNamespace | GithubNamespace
+
+export class RegistryResponse {
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.dto.ts:13]: We are not used Response and Request postfixes.

File: web/crux/src/app/registry/registry.service.ts [hunk 3]
```diff
@@ -34,11 +31,28 @@ export default class RegistryService {
...
+            images: true,
+          },
+        },
+      },
+      where: {
+        id,
+      },
+    })
+
+    return this.mapper.detailsToProto(registry)
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.service.ts:52]: `detailsToProto` the name is wrong.

File: web/crux/src/app/registry/registry.service.ts [hunk 4]
```diff
@@ -34,11 +31,28 @@ export default class RegistryService {
     })
 
     return {
-      data: registries.map(it => this.mapper.listItemToProto(it)),
+      data: registries.map(it => this.mapper.listItemToDTO(it)),
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.service.ts:34]: `listItemToDto`

File: web/crux/src/app/registry/registry.mapper.ts [hunk 5]
```diff
@@ -163,45 +146,14 @@ export default class RegistryMapper {
...
+  // TODO (@robot9706): Required by ImageMapper, remove when ImageMapper is removed
+  typeToProto(type: RegistryTypeEnum): ProtoRegistryType {
+    return protoRegistryTypeFromJSON(type.toUpperCase())
   }
 }
 
 type RegistryWithCount = Registry & {
-  _count: {
-    images: number
+  _count?: {
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.mapper.ts:156]: Why optional? If you don't need these values use the `Registry` type.

File: web/crux/src/app/registry/registry.mapper.ts [hunk 6]
```diff
@@ -1,37 +1,28 @@
...
 } from 'src/grpc/protobuf/proto/crux'
 import { REGISTRY_HUB_URL } from 'src/shared/const'
+import { CreateRegistryRequest, RegistryDetailsResponse, RegistryResponse, UpdateRegistryRequest } from './registry.dto'
 
 type RegistryTypeUnion = Pick<Registry, 'url' | 'type' | 'apiUrl' | 'user' | 'token' | 'imageNamePrefix' | 'namespace'>
 
 @Injectable()
 export default class RegistryMapper {
-  listItemToProto(registry: Registry): RegistryResponse {
+  listItemToDTO(registry: Registry): RegistryResponse {
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.mapper.ts:15]: `listItemToDto`

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 7]
```diff
@@ -1,28 +1,85 @@
...
-  @AuditLogLevel('disabled')
+  @UseGuards(JwtAuthGuard) // TODO(@robot9706): Move to controller when crux.ts getRegistryDetails is fixed
+  @AuditLogLevel('disabled') // TODO(@robot9706): Refactor the auditlog after removing gRPC
   async getRegistries(@IdentityFromRequest() identity: Identity): Promise<RegistryListResponse> {
     return await this.service.getRegistries(identity)
   }
+
+  @Get('/:id')
+  // TODO(@robot9706): Refactor the auditlog after removing gRPC
+  // TODO (robot9706): Access check?
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:43]: minor: wrong comment format

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 8]
```diff
@@ -1,28 +1,85 @@
...
+    @IdentityFromRequest() identity: Identity,
+    @Response() res: Res,
+  ): Promise<void> {
+    await this.service.updateRegistry(id, request, identity)
+
+    res.status(204).end()
+  }
+
+  @Delete('/:id')
+  @UseGuards(JwtAuthGuard) // TODO(@robot9706): Move to controller when crux.ts getRegistryDetails is fixed
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:78]: What is the main problem with the `getRegistryDetails` function?

File: web/crux/src/swagger/crux.dto.ts [hunk 9]
```diff
@@ -32,14 +32,6 @@ import {
   VersionResponse,
```
Comment (polaroi8d) [inline web/crux/src/swagger/crux.dto.ts:32]: Okey for deleting the unused values from here, but we'll delete the whole file.

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 10]
```diff
@@ -1,28 +1,85 @@
...
+  @AuditLogLevel('disabled')
+  async createRegistry(
+    @Body() request: CreateRegistryRequest,
+    @IdentityFromRequest() identity: Identity,
+    @Response() res: Res,
+  ): Promise<void> {
+    const registry = await this.service.createRegistry(request, identity)
+
+    res.set('Location', `/registries/${registry.id}`).status(201).json(registry)
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:59]: insted of `set()` you can use `res.location(`...`)`, no?

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 11]
```diff
@@ -1,28 +1,85 @@
...
+  }
+
+  @Put('/:id')
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:62]: When I tried this, works without the slash. eg.: `@Put(':id')`

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 12]
```diff
@@ -1,28 +1,85 @@
...
+  @UseGuards(JwtAuthGuard) // TODO(@robot9706): Move to controller when crux.ts getRegistryDetails is fixed
+  @AuditLogLevel('disabled')
+  @UseInterceptors(UpdateRegistryInterceptor)
+  async updateRegistry(
+    @Param('id') id: string,
+    @Body() request: UpdateRegistryRequest,
+    @IdentityFromRequest() identity: Identity,
+    @Response() res: Res,
+  ): Promise<void> {
+    await this.service.updateRegistry(id, request, identity)
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:72]: When updating a resource we wan't do add the location too, no? Any opinion @m8vago ?

File: web/crux/src/app/registry/registry.dto.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+  description?: string
+
+  icon?: string
+
+  hub?: HubRegistryDetails
+
+  v2?: V2RegistryDetails
+
+  gitlab?: GitlabRegistryDetails
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.dto.ts:124]: These works with Swagger without `@ApiProperty`? Did you check Swagger?

**Review by polaroi8d — CHANGES_REQUESTED** [review]

File: web/crux/src/app/registry/registry.service.ts (hunk 3 above)
Comment (robot9706) [reply web/crux/src/app/registry/registry.service.ts:52]: True, oops 🤔

File: web/crux/src/swagger/crux.dto.ts (hunk 9 above)
Comment (robot9706) [reply web/crux/src/swagger/crux.dto.ts:32]: I know, I just wanted to clean it up a bit.

File: web/crux/src/app/registry/registry.http.controller.ts (hunk 8 above)
Comment (robot9706) [reply web/crux/src/app/registry/registry.http.controller.ts:78]: I just realised `WsConnection` has the cookie I need, so I can add auth guards back to the whole controller.

File: web/crux/src/app/registry/registry.http.controller.ts (hunk 12 above)
Comment (robot9706) [reply web/crux/src/app/registry/registry.http.controller.ts:72]: I tought about it too. The article m8 linked in the ticket did not say anything about this though.

File: web/crux/src/swagger/crux.dto.ts (hunk 9 above)
Comment (polaroi8d) [reply web/crux/src/swagger/crux.dto.ts:32]: Okey.

File: web/crux/src/app/registry/registry.http.controller.ts (hunk 12 above)
Comment (polaroi8d) [reply web/crux/src/app/registry/registry.http.controller.ts:72]: Talk about it in person with the team.

Comment (robot9706) [reply web/crux/src/app/registry/registry.http.controller.ts:70]: Leaving it like this, as we've discussed in person.

File: web/crux/src/app/registry/registry.dto.ts (hunk 13 above)
Comment (robot9706) [reply web/crux/src/app/registry/registry.dto.ts:134]: Yes they show up with all of their fields.

Comment (polaroi8d) [reply web/crux/src/app/registry/registry.dto.ts:134]: Good to hear, comment resolved.

**Pull request merged**